shaper.go

```diff
@@ -1,6 +1,141 @@
 package shaping
 
+import (
+	"fmt"
+
+	"github.com/benoitkugler/textlayout/harfbuzz"
+	"github.com/go-text/di"
+	"golang.org/x/image/font/sfnt"
+	"golang.org/x/image/math/fixed"
+)
+
 type Shaper interface {
 	// Shape takes an Input and shapes it into the Output.
 	Shape(Input) Output
 }
+
+// scale returns x divided by unitsPerEm, rounded to the nearest fixed.Int26_6
+// value (1/64th of a pixel). Borrowed from sfnt package.
+func scale(x fixed.Int26_6, unitsPerEm sfnt.Units) fixed.Int26_6 {
+	if x >= 0 {
+		x += fixed.Int26_6(unitsPerEm) / 2
+	} else {
+		x -= fixed.Int26_6(unitsPerEm) / 2
+	}
+	return x / fixed.Int26_6(unitsPerEm)
+}
+
+// scalePosition using specific ppem and upem values.
+func scalePosition(gp harfbuzz.GlyphPosition, ppem fixed.Int26_6, upem sfnt.Units) harfbuzz.GlyphPosition {
+	gp.XAdvance = int32(scale(fixed.I(int(gp.XAdvance)).Mul(ppem), upem).Round())
+	gp.YAdvance = int32(scale(fixed.I(int(gp.YAdvance)).Mul(ppem), upem).Round())
+	gp.XOffset = int32(scale(fixed.I(int(gp.XOffset)).Mul(ppem), upem).Round())
+	gp.YOffset = int32(scale(fixed.I(int(gp.YOffset)).Mul(ppem), upem).Round())
+	return gp
+}
+
+// TODO: can we just make Input and Output into structs?
+// Interfaces encapsulate varying behavior well, but these
+// types have no behavior. They're just structured data.
+// It feels silly to wrap them in accessors like this.
+type output struct {
+	DotAdvance    fixed.Int26_6
+	TopToBaseline fixed.Int26_6
+	TextBounds    fixed.Rectangle26_6
+	Glyphs        []Glyph
+}
+
+func (o output) Advance() fixed.Int26_6 {
+	return o.DotAdvance
+}
+
+func (o output) Baseline() fixed.Int26_6 {
+	return o.TopToBaseline
+}
+
+func (o output) Bounds() fixed.Rectangle26_6 {
+	return o.TextBounds
+}
+
+func (o output) Length() int {
+	return len(o.Glyphs)
+}
+
+func (o output) Index(i int) Glyph {
+	return o.Glyphs[i]
+}
+
+// Shape turns an input into an output.
+func Shape(input Input) Output {
+	// Prepare to shape the text.
+	// TODO: maybe reuse these buffers for performance?
+	buf := harfbuzz.NewBuffer()
+	runes, start, end := input.Text()
+	buf.AddRunes(runes, start, end-start)
+	// TODO: handle vertical text?
+	switch input.Direction() {
+	case di.DirectionLTR:
+		buf.Props.Direction = harfbuzz.LeftToRight
+	case di.DirectionRTL:
+		buf.Props.Direction = harfbuzz.RightToLeft
+	}
+	buf.Props.Language = input.Language()
+	buf.Props.Script = input.Script()
+	// TODO: figure out what (if anything) to do if this type assertion fails.
+	font := harfbuzz.NewFont(input.Face().(harfbuzz.Face))
+	upem := sfnt.Units(font.Face().Upem())
+	ppem := input.Size()
+
+	// Actually use harfbuzz to shape the text.
```

Comment:
We have opposite expectations. Since @benoitkugler has already gone through the trouble of porting harfbuzz to Go, I'd like to reuse that work rather than reimplementing it. In case it's the name that is causing confusion, this is leveraging the Go port, not the C++ codebase.

My perception of this project was to develop common code for accessing shaping as well as higher-level things like bidi splitting, paragraph layout, and font management.

Reply:
Ack, makes sense

Reply:
I would suggest that, at long term, we aim at moving the implementation from `github.com/benoitkugler/textlayout` to this repository. 
I think it would be valuable for two reasons :
- probably better code hygiene on a shared, "official" repository 
- the scope of the shaper, as it is now, is very close to (almost directly copied from) what harfbuzz is doing, so that having two packages for the same task feels a bit odd.

However, since it's not an easy task it seems fine to start by just importing `github.com/benoitkugler/textlayout` and using it. 
When we get a better understanding of the process, we could then think of merging.

Reply:
In the same idea, concerning the `harfbuzz.Face` interface, notice that [this PR](https://github.com/go-text/font/pull/1) is a first step towards our own `Face`, with the long term goal to integrate font management into this project.